core/component_manager.ts

```diff
@@ -195,12 +195,12 @@ export class ComponentManager {
         return a.weight - b.weight;
       });
       componentDataList.forEach(function(ComponentDatum) {
-        components.push(ComponentDatum.component);
+        components.push(ComponentDatum.component as T);
```

Comment:
TS infers the type as ComponentDatum from the typing of `componentDataList`. If you're referring to the "as T" cast I think we're stuck with that because, while `capabilityToComponentIds_` should be giving us only ComponentDatums whose component field is of type T, TS doesn't know that.

Reply:
I think I don't fully understand the purpose of the `ComponentDatum` type here, but the code as revised in this PR looks plausible enough so don't let that stop you submitting it!